diff --git a/packages/react-dom/src/__tests__/ReactDOMSingletonComponents-test.js b/packages/react-dom/src/__tests__/ReactDOMSingletonComponents-test.js
--- a/packages/react-dom/src/__tests__/ReactDOMSingletonComponents-test.js
+++ b/packages/react-dom/src/__tests__/ReactDOMSingletonComponents-test.js
@@ -402,6 +402,31 @@
     expect(body.innerHTML).toBe('<span>managed child</span>');
   });
 
+  it('replaces dangerouslySetInnerHTML with children in head', async () => {
+    const root = ReactDOMClient.createRoot(document);
+    const head = document.head;
+
+    root.render(
+      <html>
+        <head
+          dangerouslySetInnerHTML={{__html: '<meta name="managed-stale" />'}}
+        />
+        <body />
+      </html>,
+    );
+    await waitForAll([]);
+    expect(head.innerHTML).toBe('<meta name="managed-stale">');
+
+    root.render(
+      <html>
+        <head>
+          <title>managed title</title>
+        </head>
+        <body />
+      </html>,
+    );
+    await waitForAll([]);
+
+    expect(head.innerHTML).toBe('<title>managed title</title>');
+  });
+
   // @gate TODO
   it('preserves imperative attributes when acquiring a singleton', async () => {
     const body = document.body;
