diff --git a/packages/react-reconciler/src/ReactFiberBeginWork.js b/packages/react-reconciler/src/ReactFiberBeginWork.js
--- a/packages/react-reconciler/src/ReactFiberBeginWork.js
+++ b/packages/react-reconciler/src/ReactFiberBeginWork.js
@@ -2090,7 +2090,7 @@
 
   if (
-    workInProgress.type !== 'html' &&
+    workInProgress.type === 'body' &&
     prevProps !== null &&
     shouldSetTextContent(workInProgress.type, prevProps) &&
     !shouldSetTextContent(workInProgress.type, nextProps)
